### PR TITLE
Add a read-optimized implementation of relational metadata.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Part of #8258

### CODE REVIEW

File: src/EFCore/Metadata/IReadOnlyServiceProperty.cs [hunk 1]
```diff
@@ -22,6 +23,17 @@ public interface IReadOnlyServiceProperty : IReadOnlyPropertyBase
...
         ServiceParameterBinding? ParameterBinding { get; }
 
+        /// <summary>
+        ///     <para>
+        ///         Gets the <see cref="PropertyAccessMode" /> being used for this property.
+        ///         <see langword="null" /> indicates that the default property access mode is being used.
+        ///     </para>
+        /// </summary>
+        /// <returns> The access mode being used. </returns>
+        PropertyAccessMode IReadOnlyPropertyBase.GetPropertyAccessMode()
```
Comment (smitpatel) [inline src/EFCore/Metadata/IReadOnlyServiceProperty.cs:33]: Why does this DIM exist here but not on IReadOnlyProperty?

File: test/EFCore.Tests/Metadata/Conventions/ConventionDispatcherTest.cs [hunk 2]
```diff
@@ -1926,7 +1926,7 @@ public void OnNavigationRemoved_calls_conventions_in_order(bool useBuilder, bool
             }
             else
             {
-                Assert.NotNull(relationshipBuilder.Metadata.SetDependentToPrincipal((string)null, ConfigurationSource.Convention));
+                relationshipBuilder.Metadata.SetDependentToPrincipal((string)null, ConfigurationSource.Convention);
```
Comment (smitpatel) [inline test/EFCore.Tests/Metadata/Conventions/ConventionDispatcherTest.cs:1929]: if verifies some assert else does not.

**Review by smitpatel — APPROVED** [review]
:shipit:

File: src/EFCore/Metadata/IReadOnlyServiceProperty.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/IReadOnlyServiceProperty.cs:33]: Will push down to class

File: test/EFCore.Tests/Metadata/Conventions/ConventionDispatcherTest.cs (hunk 2 above)
Comment (AndriySvyryd) [reply test/EFCore.Tests/Metadata/Conventions/ConventionDispatcherTest.cs:1929]: I'll make the Assert smarter

**Pull request merged**